toqito/matrix_ops/tensor.py

```diff
@@ -135,15 +135,22 @@ def tensor(*args) -> np.ndarray:
     # Tensor product one matrix `n` times with itself.
     if len(args) == 2 and isinstance(args[1], int):
         num_tensor = args[1]
+        if num_tensor == 0:
+            return None
         if num_tensor == 1:
             return args[0]
-        if num_tensor == 2:
-            return np.kron(args[0], args[0])
-        if num_tensor >= 3:
-            result = np.kron(args[0], args[0])
-            for _ in range(2, num_tensor):
-                result = np.kron(result, args[0])
-        return result
+        mat = args[0]
+        tensored_twice = np.kron(mat, mat)
+
+        def fast_exp(matrix, q):
+            if q == 1:
+                return matrix
+            tmp = fast_exp(tensored_twice, q >> 1)
+            if q & 1:
+                return np.kron(mat, tmp)
+            return tmp
+
+        return fast_exp(mat, num_tensor)
 
     # Tensor product between two or more matrices.
```

Comment:
Nice PR, @tnemoz ! 

How about considering the following layout:

```py
import numpy as np

def tensor(*args) -> np.ndarray:
    r"""Compute the Kronecker tensor product."""

    def fast_exp(matrix, q):
        """Efficient exponentiation by squaring."""
        if q == 1:
            return matrix
        tmp = fast_exp(matrix, q >> 1)
        tmp = np.kron(tmp, tmp)
        if q & 1:  # If q is odd
            tmp = np.kron(matrix, tmp)
        return tmp

    result = None

    # Input is provided as a list of numpy matrices.
    if (len(args) == 1 and isinstance(args[0], list)) or (len(args) == 1 and isinstance(args[0], np.ndarray)):
        if len(args[0]) == 1:
            return args[0][0]
        if len(args[0]) == 2:
            return np.kron(args[0][0], args[0][1])
        if len(args[0]) >= 3:
            result = args[0][0]
            for i in range(1, len(args[0])):
                result = np.kron(result, args[0][i])
        return result

    # Tensor product one matrix `n` times with itself.
    if len(args) == 2 and isinstance(args[1], int):
        num_tensor = args[1]
        if num_tensor == 0:
            return np.eye(args[0].shape[0])
        if num_tensor == 1:
            return args[0]
        return fast_exp(args[0], num_tensor)

    # Tensor product between two or more matrices.
    if len(args) == 2:
        return np.kron(args[0], args[1])
    if len(args) >= 3:
        result = args[0]
        for i in range(1, len(args)):
            result = np.kron(result, args[i])
        return result

    raise ValueError("The `tensor` function must take either a matrix or vector.")
```

Basically the main change(s) here are:

1. Moving the `fast_exp` function to the top of the `tensor` function to make it a bit more readable and less nested.
2. The updated fast_exp avoids unnecessary recomputation of the tensor product in the recursive step.
3. For n = 0, we now return the identity matrix, which aligns with the standard interpretation of a tensor product raised to the power of 0.

Any thoughts on this change?

Reply:
> Nice PR

It isn't, I'm horrified to look back at the code I wrote :') I'm even a bit surprised it passes the test, I think I can add one that won't pass, there's a mistake in my code that isn't present in yours. Sorry for that! I'll modify ASAP.

Concerning the n=0 case, I'm all for returning something more canonic than None. However, I would've argued that the interpretation would have been to return `[[1]]`, since this is the neutral element for tensor products. In particular, by returning this, we keep the additivity property in the powers: $a^{\otimes n}\otimes a^{\otimes0}=a^{\otimes (n+0)}$, which wouldn't work with the identity  Or is there another interpretation that I don't know of?

Reply:
> It isn't, I'm horrified to look back at the code I wrote :') 

Completely disagree. In fact, I sincerely appreciate all of your efforts to contribute to |toqito>. On that note, thank you again for taking the time to add your code, and please be kinder to yourself!

> Concerning the n=0 case...

Yes, definitely. I think returning `None` is actually a bug and that `[[1]]` makes more sense. Making the change to reflect that would make sense if you are able to do so!

Thank you again for your efforts!